06/Coordinatron.scala

```diff
@@ -1,129 +1,119 @@
-      if (p._2 || l.contains(p)) l else p._1 :: l
+    // adds a point to the list if a) it's not infinite and b) it's not already in the list (no duplicates!)
+    def updateList(l: List[Point], p: MapPoint): List[Point] = {
+      if (p.isInfinite || l.contains(p)) l else p.point :: l
     }
 
-    def go(l: List[(Int, Int)], bounds: (Int, Int, Int, Int), x: Int, y: Int, m: Map[(Int, Int), Int], infinites: List[(Int, Int)]): (Map[(Int, Int), Int], List[(Int, Int)]) = {
-      if (y > bounds._4 + 2) (m, infinites)
+    def go(x: Int, y: Int, m: Map[Point, Int], infinites: List[Point]): (Map[Point, Int], List[Point]) = {
+      if (y > bounds.maxY + 2) (m, infinites)
       else {
-        val currentX: Int = if (x > bounds._2 + 1) bounds._1 - 1 else x
+        val currentX: Int = if (x > bounds.maxX + 1) bounds.minX - 1 else x
         val currentY: Int = if (currentX < x) y + 1 else y
-        val closestPoint: ((Int, Int), Boolean) = findClosestPoint((currentX, currentY), l)
-        val is: List[(Int, Int)] = if (x < bounds._1 || x > bounds._2 || y < bounds._3 || y > bounds._4) updateList(infinites, closestPoint) else infinites
-        go(l, bounds, currentX + 1, currentY, updateMap(m, closestPoint), is)
+        val closestPoint: MapPoint = findClosestPoint(Point(currentX, currentY), l)
+        val infinite: List[Point] = if (x < bounds.minX || x > bounds.maxX || y < bounds.minY || y > bounds.maxY) updateList(infinites, closestPoint) else infinites
+        go(currentX + 1, currentY, updateMap(m, closestPoint), infinite)
       }
     }
 
-    go(l, bounds, bounds._1 - 1, bounds._3 - 1, Map(), List())
+    go(bounds.minX - 1, bounds.minY - 1, Map(), List())
   }
 
-  def findGreatestArea(m: Map[(Int, Int), Int], is: List[(Int, Int)]): Int = {
-    def go(m: Map[(Int, Int), Int], is: List[(Int, Int)], acc: Int): Int = {
+  def findGreatestArea(m: Map[Point, Int], infinites: List[Point]): Int = {
+    def go(m: Map[Point, Int], acc: Int): Int = {
       if (m.isEmpty) acc
-      else go(m - m.head._1, is, if (is.contains(m.head._1) || max(m.head._2, acc) == acc) acc else m.head._2)
+      else go(m - m.head._1, if (infinites.contains(m.head._1) || m.head._2.max(acc) == acc) acc else m.head._2)
     }
-    go(m, is, 0)
+    go(m, 0)
   }
 
-  def findSafestCoordinates(ps: List[(Int, Int)], bounds: (Int, Int, Int, Int)): Int = {
-    def sumDistances(p: (Int, Int), ps: List[(Int, Int)]): Int = {
-      def go(p: (Int, Int), ps: List[(Int, Int)], acc: Int): Int = {
+  def findSafestCoordinates(points: List[Point], bounds: Bounds): Int = {
+    def sumDistances(p: Point, ps: List[Point]): Int = {
+      def go(ps: List[Point], acc: Int): Int = {
         if (ps.isEmpty || acc > 10000) acc
-        else go(p, ps.tail, acc + distance(p, ps.head))
+        else go(ps.tail, acc + distance(p, ps.head))
       }
-      println(s"Summing distances for ${p}")
-      go(p, ps, 0)
+      go(points, 0)
```

Comment:
You can often find combinators that are easier to use than recursion: `points.map(distance(p, _)). sum`